packages/devtools_app/lib/src/memory/flutter/memory_chart.dart

```diff
@@ -222,63 +203,18 @@ class MemoryChartState extends State<MemoryChart> with AutoDisposeMixin {
   // Trace #3 External Memory used.
   LineDataSet externalMemorySet;
 
-  /// Common utility function to handle loading of the data into the
-  /// chart for either offline or live Feed.
-  void _processData(List<HeapSample> data, int startingDataIndex) {
-    for (var dataIndex = startingDataIndex;
-        dataIndex < data.length;
-        dataIndex++) {
-      final sample = data[dataIndex];
-      final timestamp = sample.timestamp.toDouble();
-
-      final capacity = sample.capacity.toDouble();
-      final used = sample.used.toDouble();
-      final external = sample.external.toDouble();
-
-      final extEntry = Entry(
-        x: timestamp,
-        y: external,
-        icon: _img,
-      );
-      final usedEntry = Entry(
-        x: timestamp,
-        y: used + external,
-        icon: _img,
-      );
-      final capacityEntry = Entry(
-        x: timestamp,
-        y: capacity,
-        icon: _img,
-      );
-
-      setState(() {
-        _externalHeap.add(extEntry);
-        _used.add(usedEntry);
-        _capacity.add(capacityEntry);
-      });
-    }
-
-    updateChart();
-  }
-
-  /// Fetch all the data in the loaded from a memory log (JSON file in /tmp).
-  void processMemoryLogFileData() {
-    assert(controller.offline);
-    assert(memoryTimeline.offflineData.isNotEmpty);
-    _processData(memoryTimeline.offflineData, 0);
-  }
-
-  void processLiveData([bool reloadAllData = false]) {
-    assert(!controller.offline);
-    assert(memoryTimeline.data.isNotEmpty);
-
-    final List<HeapSample> liveFeed = memoryTimeline.data;
-    if (_used.length != liveFeed.length || reloadAllData) {
-      _processData(liveFeed, _used.length);
-    }
+  /// General utility function handles loading all heap samples (online or offline) and
+  /// or the latest
+  /// heap sample.
+  void _processAndUpdate([bool reloadAllData = false]) {
+    setState(() {
+      controller.processData(reloadAllData);
+      _updateChart();
+    });
   }
 
-  void updateChart() {
+  /// Display any newly received heap sample(s) in the chart.
+  void _updateChart() {
     setState(() {
       // Signal data has changed.
       usedHeapSet.notifyDataSetChanged();
```

Comment:
instead of 
`chartController.data = LineData.fromList(
          []..add(usedHeapSet)..add(externalMemorySet)..add(capacityHeapSet));`
below, you can do
`chartController.data = LineData.fromList([...usedHeapSet, ...externalMemorySet, ...capacityHeapSet]);`